Source/Initialization/WarpXInitData.cpp

```diff
@@ -1072,48 +1059,61 @@ WarpX::InitLevelData (int lev, Real /*time*/)
-       [[maybe_unused]] const char field,
-       const int lev, PatchType patch_type)
+void WarpX::ComputeExternalFieldOnGridUsingParser (
+    warpx::fields::FieldType field,
+    amrex::ParserExecutor<4> const& fx_parser,
+    amrex::ParserExecutor<4> const& fy_parser,
+    amrex::ParserExecutor<4> const& fz_parser,
+    int lev, PatchType patch_type, [[maybe_unused]] const char topology,
+    std::optional<ablastr::fields::VectorField> const& edge_lengths,
+    std::optional<ablastr::fields::VectorField> const& face_areas)
 {
+    auto t = gett_new(lev);
 
     auto dx_lev = geom[lev].CellSizeArray();
-    amrex::IntVect refratio = (lev > 0 ) ? WarpX::RefRatio(lev-1) : amrex::IntVect(1);
+    const RealBox& real_box = geom[lev].ProbDomain();
+
+    amrex::IntVect refratio = (lev > 0 ) ? RefRatio(lev-1) : amrex::IntVect(1);
     if (patch_type == PatchType::coarse) {
         for (int idim = 0; idim < AMREX_SPACEDIM; ++idim) {
-            dx_lev[idim] = dx_lev[idim] * refratio[idim];
+            dx_lev[idim] = dx_lev[idim] + refratio[idim];
         }
     }
-    const RealBox& real_box = geom[lev].ProbDomain();
+
+    using ablastr::fields::Direction;
+    amrex::MultiFab* mfx = m_fields.get(field, Direction{0}, lev);
+    amrex::MultiFab* mfy = m_fields.get(field, Direction{1}, lev);
+    amrex::MultiFab* mfz = m_fields.get(field, Direction{2}, lev);
+
     const amrex::IntVect x_nodal_flag = mfx->ixType().toIntVect();
     const amrex::IntVect y_nodal_flag = mfy->ixType().toIntVect();
     const amrex::IntVect z_nodal_flag = mfz->ixType().toIntVect();
 
-    bool const eb_enabled = EB::enabled();
+    const bool eb_enabled = EB::enabled();
 
     for ( MFIter mfi(*mfx, TilingIfNotGPU()); mfi.isValid(); ++mfi) {
-        const amrex::Box &tbx = mfi.tilebox(x_nodal_flag, mfx->nGrowVect());
-        const amrex::Box &tby = mfi.tilebox(y_nodal_flag, mfy->nGrowVect());
-        const amrex::Box &tbz = mfi.tilebox(z_nodal_flag, mfz->nGrowVect());
+        const amrex::Box& tbx = mfi.tilebox( x_nodal_flag, mfx->nGrowVect() );
+        const amrex::Box& tby = mfi.tilebox( y_nodal_flag, mfy->nGrowVect() );
+        const amrex::Box& tbz = mfi.tilebox( z_nodal_flag, mfz->nGrowVect() );
 
-        auto const &mfxfab = mfx->array(mfi);
-        auto const &mfyfab = mfy->array(mfi);
-        auto const &mfzfab = mfz->array(mfi);
+        auto const& mfxfab = mfx->array(mfi);
+        auto const& mfyfab = mfy->array(mfi);
+        auto const& mfzfab = mfz->array(mfi);
 
         amrex::Array4<amrex::Real> lx, ly, lz, Sx, Sy, Sz;
         if (eb_enabled) {
-            lx = edge_lengths[0]->array(mfi);
-            ly = edge_lengths[1]->array(mfi);
-            lz = edge_lengths[2]->array(mfi);
-            Sx = face_areas[0]->array(mfi);
-            Sy = face_areas[1]->array(mfi);
-            Sz = face_areas[2]->array(mfi);
+            if (edge_lengths.has_value()) {
+                auto& edge_lengths_array = edge_lengths.value();
```

Comment:
`clang-tidy` wants this to be
```suggestion
                const auto& edge_lengths_array = edge_lengths.value();
```